main.go

```diff
@@ -1,426 +0,0 @@
-		proxy.Transport = &http2.Transport{
-			// Allow http schema. This doesn't automatically disable TLS
-			AllowHTTP: true,
-			// Do disable TLS.
-			// In combination with the schema check above. We could enforce h2c against the upstream server
-			DialTLS: func(netw, addr string, cfg *tls.Config) (net.Conn, error) {
-				return net.Dial(netw, addr)
-			},
-		}
-	}
-
-	mux := http.NewServeMux()
-	mux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		found := len(cfg.allowPaths) == 0
-		for _, pathAllowed := range cfg.allowPaths {
-			found, err = path.Match(pathAllowed, req.URL.Path)
-			if err != nil {
-				return
-			}
-			if found {
-				break
-			}
-		}
-		if !found {
-			http.NotFound(w, req)
-			return
-		}
-
-		ignorePathFound := false
-		for _, pathIgnored := range cfg.ignorePaths {
-			ignorePathFound, err = path.Match(pathIgnored, req.URL.Path)
-			if err != nil {
-				return
-			}
-			if ignorePathFound {
-				break
-			}
-		}
-
-		if !ignorePathFound {
-			ok := auth.Handle(w, req)
-			if !ok {
-				return
-			}
-		}
-
-		proxy.ServeHTTP(w, req)
-	}))
-
-	var gr run.Group
-	{
-		if cfg.secureListenAddress != "" {
-			srv := &http.Server{Handler: mux, TLSConfig: &tls.Config{}}
-
-			if cfg.tls.certFile == "" && cfg.tls.keyFile == "" {
-				klog.Info("Generating self signed cert as no cert is provided")
-				host, err := os.Hostname()
-				if err != nil {
-					klog.Fatalf("Failed to retrieve hostname for self-signed cert: %v", err)
-				}
-				certBytes, keyBytes, err := certutil.GenerateSelfSignedCertKey(host, nil, nil)
-				if err != nil {
-					klog.Fatalf("Failed to generate self signed cert and key: %v", err)
-				}
-				cert, err := tls.X509KeyPair(certBytes, keyBytes)
-				if err != nil {
-					klog.Fatalf("Failed to load generated self signed cert and key: %v", err)
-				}
-
-				srv.TLSConfig.Certificates = []tls.Certificate{cert}
-			} else {
-				klog.Info("Reading certificate files")
-				ctx, cancel := context.WithCancel(context.Background())
-				r, err := rbac_proxy_tls.NewCertReloader(cfg.tls.certFile, cfg.tls.keyFile, cfg.tls.reloadInterval)
-				if err != nil {
-					klog.Fatalf("Failed to initialize certificate reloader: %v", err)
-				}
-
-				srv.TLSConfig.GetCertificate = r.GetCertificate
-
-				gr.Add(func() error {
-					return r.Watch(ctx)
-				}, func(error) {
-					cancel()
-				})
-			}
-
-			version, err := k8sapiflag.TLSVersion(cfg.tls.minVersion)
-			if err != nil {
-				klog.Fatalf("TLS version invalid: %v", err)
-			}
-
-			cipherSuiteIDs, err := k8sapiflag.TLSCipherSuites(cfg.tls.cipherSuites)
-			if err != nil {
-				klog.Fatalf("Failed to convert TLS cipher suite name to ID: %v", err)
-			}
-
-			srv.TLSConfig.CipherSuites = cipherSuiteIDs
-			srv.TLSConfig.MinVersion = version
-			srv.TLSConfig.ClientAuth = tls.RequestClientCert
-
-			if err := http2.ConfigureServer(srv, nil); err != nil {
```

Comment:
TODO, if this isn't replaced with upstream's serving library, the tls and listener logic here needs review.